File: sdk-python/gatedhouse/core/delegation/cache.py
"""Delegation cache — PostgreSQL-backed cache for Sphinx delegation data."""

from __future__ import annotations

import json
import logging
from datetime import datetime

from gatedhouse.core.types import CachedDelegation
from gatedhouse.database.connection import DatabaseConnection

logger = logging.getLogger("gatedhouse.delegation.cache")


class DelegationCache:
    """PostgreSQL-backed delegation cache synced from Sphinx events."""

    def __init__(self, db: DatabaseConnection) -> None:
        self._db = db

    async def upsert(self, delegation: CachedDelegation) -> None:
        await self._db.execute(
            """INSERT INTO gatedhouse_delegation_cache
               (delegation_id, agent_id, delegator_id, delegator_membership_id,
                org_id, scopes, constraints, max_uses, use_count, status, expires_at)
               VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)
               ON CONFLICT (delegation_id) DO UPDATE SET
               scopes = EXCLUDED.scopes,
               constraints = EXCLUDED.constraints,
               max_uses = EXCLUDED.max_uses,
               use_count = EXCLUDED.use_count,
               status = EXCLUDED.status,
               expires_at = EXCLUDED.expires_at,
               synced_at = NOW()""",
            delegation.delegation_id,
            delegation.agent_id,
            delegation.delegator_id,
            delegation.delegator_membership_id,
            delegation.org_id,
            json.dumps(delegation.scopes),
            json.dumps(delegation.constraints),
            delegation.max_uses,
            delegation.use_count,
            delegation.status,
            delegation.expires_at,
        )

    async def find_active(self, delegation_id: str) -> CachedDelegation | None:
        row = await self._db.query_one(
            """SELECT * FROM gatedhouse_delegation_cache
               WHERE delegation_id = $1 AND status = 'active'""",
            delegation_id,
        )
        return self._to_cached(row) if row else None

    async def update_status(self, delegation_id: str, status: str) -> None:
        await self._db.execute(
            "UPDATE gatedhouse_delegation_cache SET status = $2, synced_at = NOW() WHERE delegation_id = $1",
            delegation_id, status,
        )

    async def revoke_all_for_agent(self, agent_id: str) -> None:
        await self._db.execute(
            "UPDATE gatedhouse_delegation_cache SET status = 'revoked', synced_at = NOW() WHERE agent_id = $1",
            agent_id,
        )

    async def remove_all_for_org(self, org_id: str) -> None:
        await self._db.execute(
            "DELETE FROM gatedhouse_delegation_cache WHERE org_id = $1", org_id
        )

    @staticmethod
    def _to_cached(row: dict) -> CachedDelegation:
        scopes = row["scopes"]
        if isinstance(scopes, str):
            scopes = json.loads(scopes)
        constraints = row["constraints"]
        if isinstance(constraints, str):
            constraints = json.loads(constraints)
        return CachedDelegation(
            delegation_id=row["delegation_id"],
            agent_id=row["agent_id"],
            delegator_id=row["delegator_id"],
            delegator_membership_id=row["delegator_membership_id"],
            org_id=row["org_id"],
            scopes=scopes,
            constraints=constraints,
            max_uses=row["max_uses"],
            use_count=row["use_count"],
            status=row["status"],
            expires_at=row["expires_at"],
            synced_at=row.get("synced_at"),
        )

File: sdk-typescript/src/delegation/cache.ts
/**
 * Delegation cache — local cache of Sphinx delegation data.
 *
 * Delegation data is synced via Sphinx events and cached in the
 * service's own database for fast lookups during delegated agent requests.
 */

import { DatabaseConnection } from '../database/connection';
import { CachedDelegation } from '../types';
import { createLogger } from '../logger';

const logger = createLogger('delegation-cache');

interface DelegationRow {
  delegation_id: string;
  agent_id: string;
  delegator_id: string;
  delegator_membership_id: string;
  org_id: string;
  scopes: string[];
  constraints: Record<string, unknown>;
  max_uses: number | null;
  use_count: number;
  status: string;
  expires_at: Date;
  synced_at: Date;
}

function toCachedDelegation(row: DelegationRow): CachedDelegation {
  return {
    delegationId: row.delegation_id,
    agentId: row.agent_id,
    delegatorId: row.delegator_id,
    delegatorMembershipId: row.delegator_membership_id,
    orgId: row.org_id,
    scopes: row.scopes,
    constraints: row.constraints,
    maxUses: row.max_uses,
    useCount: row.use_count,
    status: row.status,
    expiresAt: row.expires_at,
    syncedAt: row.synced_at,
  };
}

export class DelegationCache {
  constructor(private db: DatabaseConnection) {}

  /**
   * Upsert a delegation into the local cache.
   */
  async upsert(delegation: Omit<CachedDelegation, 'syncedAt'>): Promise<void> {
    await this.db.execute(
      `INSERT INTO gatedhouse_delegation_cache
         (delegation_id, agent_id, delegator_id, delegator_membership_id,
          org_id, scopes, constraints, max_uses, use_count, status, expires_at, synced_at)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, now())
       ON CONFLICT (delegation_id) DO UPDATE SET
         agent_id = EXCLUDED.agent_id,
         delegator_id = EXCLUDED.delegator_id,
         delegator_membership_id = EXCLUDED.delegator_membership_id,
         org_id = EXCLUDED.org_id,
         scopes = EXCLUDED.scopes,
         constraints = EXCLUDED.constraints,
         max_uses = EXCLUDED.max_uses,
         use_count = EXCLUDED.use_count,
         status = EXCLUDED.status,
         expires_at = EXCLUDED.expires_at,
         synced_at = now()`,
      [
        delegation.delegationId,
        delegation.agentId,
        delegation.delegatorId,
        delegation.delegatorMembershipId,
        delegation.orgId,
        delegation.scopes,
        JSON.stringify(delegation.constraints),
        delegation.maxUses,
        delegation.useCount,
        delegation.status,
        delegation.expiresAt,
      ],
    );
  }

  /**
   * Find an active delegation for an agent.
   */
  async findActiveForAgent(
    agentId: string,
    orgId: string,
  ): Promise<CachedDelegation | null> {
    const row = await this.db.queryOne<DelegationRow>(
      `SELECT * FROM gatedhouse_delegation_cache
       WHERE agent_id = $1 AND org_id = $2 AND status = 'active'
         AND expires_at > now()
         AND (max_uses IS NULL OR use_count < max_uses)
       ORDER BY synced_at DESC
       LIMIT 1`,
      [agentId, orgId],
    );
    return row ? toCachedDelegation(row) : null;
  }

  /**
   * Find a delegation by ID.
   */
  async findById(delegationId: string): Promise<CachedDelegation | null> {
    const row = await this.db.queryOne<DelegationRow>(
      'SELECT * FROM gatedhouse_delegation_cache WHERE delegation_id = $1',
      [delegationId],
    );
    return row ? toCachedDelegation(row) : null;
  }

  /**
   * Update delegation status.
   */
  async updateStatus(delegationId: string, status: string): Promise<void> {
    await this.db.execute(
      `UPDATE gatedhouse_delegation_cache
       SET status = $1, synced_at = now()
       WHERE delegation_id = $2`,
      [status, delegationId],
    );
    logger.info({ delegationId, status }, 'Delegation status updated');
  }

  /**
   * Increment use count for a delegation.
   */
  async incrementUseCount(delegationId: string): Promise<void> {
    await this.db.execute(
      `UPDATE gatedhouse_delegation_cache
       SET use_count = use_count + 1, synced_at = now()
       WHERE delegation_id = $1`,
      [delegationId],
    );
  }

  /**
   * Revoke all delegations for an agent.
   */
  async revokeAllForAgent(agentId: string): Promise<void> {
    await this.db.execute(
      `UPDATE gatedhouse_delegation_cache
       SET status = 'revoked', synced_at = now()
       WHERE agent_id = $1 AND status = 'active'`,
      [agentId],
    );
    logger.info({ agentId }, 'All delegations revoked for agent');
  }

  /**
   * Remove all delegations for an org.
   */
  async removeAllForOrg(orgId: string): Promise<void> {
    await this.db.execute(
      'DELETE FROM gatedhouse_delegation_cache WHERE org_id = $1',
      [orgId],
    );
  }
}
